### PR TITLE
IGNITE-13200 SQL create index on invalid data type

### PR DESCRIPTION
**Pull request by tledkov-gridgain**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java [hunk 1]
```diff
@@ -2433,7 +2434,7 @@ private T doPut(T row, boolean needOld) throws IgniteCheckedException {
                 }
             }
         }
-        catch (IgniteCheckedException e) {
+        catch (IgniteCheckedException | IgniteSQLException e) {
```
Comment (ibessonov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java:2437]: Can you please remove this new exception from here? I believe that it's unused and left from previous fix attempt. Thank you!

File: modules/indexing/src/test/java/org/apache/ignite/internal/processors/query/CreateIndexOnInvalidDataTypeTest.java [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+import org.apache.ignite.configuration.IgniteConfiguration;
+import org.apache.ignite.failure.StopNodeFailureHandler;
+import org.apache.ignite.internal.processors.cache.CachePartialUpdateCheckedException;
+import org.apache.ignite.internal.processors.cache.index.AbstractIndexingCommonTest;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.testframework.GridTestUtils;
+import org.junit.Test;
+
+/**
+ * Tests for local query execution in lazy mode.
```
Comment (korlov42) [inline modules/indexing/src/test/java/org/apache/ignite/internal/processors/query/CreateIndexOnInvalidDataTypeTest.java:43]: Wrong description

File: modules/indexing/src/test/java/org/apache/ignite/internal/processors/query/CreateIndexOnInvalidDataTypeTest.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+        grid().cluster().state(ClusterState.ACTIVE);
+
+        IgniteCache<Integer, Value> c = grid().createCache(new CacheConfiguration<Integer, Value>()
+            .setName("test")
+            .setSqlSchema("PUBLIC")
+            .setQueryEntities(Collections.singleton(new QueryEntity(Integer.class, Value.class)
+                    .setTableName("TEST")
+                )
+            )
```
Comment (korlov42) [inline modules/indexing/src/test/java/org/apache/ignite/internal/processors/query/CreateIndexOnInvalidDataTypeTest.java:95]:
```
.setQueryEntities(
    Collections.singleton(
        new QueryEntity(Integer.class, Value.class).setTableName("TEST")
    )
)
```
or
```
.setQueryEntities(Collections.singleton(new QueryEntity(Integer.class, Value.class)
    .setTableName("TEST")))
```
or
```
.setQueryEntities(Collections.singleton(
    new QueryEntity(Integer.class, Value.class).setTableName("TEST")
))
```
but not mix of them

File: modules/indexing/src/test/java/org/apache/ignite/internal/processors/query/CreateIndexOnInvalidDataTypeTest.java [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+        bob.setField("VAL_DATE", new java.util.Date());
+
+        GridTestUtils.assertThrowsAnyCause(log, () -> {
+                grid().cache("test").put(0, bob.build());
+
+                return null;
+            },
+            CachePartialUpdateCheckedException.class, "Failed to update keys");
+
+        sql("DROP INDEX TEST_0_VAL_DATE_IDX");
```
Comment (korlov42) [inline modules/indexing/src/test/java/org/apache/ignite/internal/processors/query/CreateIndexOnInvalidDataTypeTest.java:161]: let's verify that after index was dropped you could insert values without any error

**Pull request closed without merging**